### ISSUE TITLE
Sitemap creation missing last category

### ISSUE THREAD

**Issue by Assambra** [magento/magento2#34797]
### Preconditions (*)

Magento Commerce 2.4.3-p1, fresh installation and my production Site
Missing 3th subcategory in sitemap. It generates a lot of trouble with SEO Canonical tag shows on -> https://www.example.com/Subcategory1/Subcategory2/Subcategory3/Product1 and Sitemap generates https://www.example.com/Subcategory1/Subcategory2/Product1

Edit:
I don't know if that has anything to do with it, but if the URL rewrite ID plays a role. they are created incorrectly from the order. I am trying to explain
id 1 https://www.example.com/Product1
id 2 https://www.example.com/Subcategory1/Subcategory2/Subcategory3/Product1
id 3 https://www.example.com/Subcategory1/Product1
id 4 https://www.example.com/Subcategory1/Subcategory2/Product1 <--It looks like the last ID are taken to generate the Sitemap

Ok it have to do with the url-rewrite ID i tryed by my self. Modify Product1 add it to Subcategory1 save, add to Subcategory2 save, add to Subcatagory3 save. The result is right order of url-rewrite ID and Sitmap works as expected and shows https://www.example.com/Subcategory1/Subcategory2/Subcategory3/Product1

It dosnt work if you select only the Subcategory3 or all Subcategories at the same time for the Product1

How can i fix that i got a lot of warning in Search console Duplicate - Submitted URL not set as canonical for this sitmap and i cannot fix 100 or more products by hand as i descriped earlyer.

### Steps to reproduce (*)

Stores->Configuraton-Catalog->Catalog->Search Engine optimization
1. Set Use Categories Path for Product URLs -> Yes
2. Catalog-->Categories -> Create Three categories each as Subcategory from the last one.
 Default Categorie
     |_>Subcategory1
         |_-->Subcategory2
             |_--->Subcategory3

3. Catalog-->Products -> Create a product and name it Product1,  set the Product Category to Subcategory3

4. Marketing->Site Map -> Create a new sitemap and generate it

### Expected result (*)

<loc>
https://www.example.com/Subcategory1/Subcategory2/Subcategory3/Product1
</loc>

### Actual result (*)

<loc>
https://www.eample.com/Subcategory1/Subcategory2/Product1
</loc>

**Additional Information:**
2.4-develop behavior: Category path is not included for product url in sitemap no matter even if all categories are assigned to the product:

![image](https://user-images.githubusercontent.com/60198231/147939735-2001a270-5199-4cd0-8aec-b1de2f5c3b56.png)

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.4/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [x] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._

**Comment by engcom-November** [comment]
@magento give me 2.4-develop instance

**Comment by engcom-November** [comment]
Verified the issue on Magento 2.4-develop branch but cannot able to reproduce the issue:
Generated sitemap after setting the Product Category to Subcategory3
http://magento24.local/default/subcategory1/subcategory2/subcategory3.html
![image](https://user-images.githubusercontent.com/60198231/147566839-195e8c9f-6d47-49e3-a760-63e7ca94af97.png)
sitemap.xml:
![image](https://user-images.githubusercontent.com/60198231/147566964-cf6fdd6a-eddb-44e0-b8b4-de6ff09a816e.png)
Kindly recheck the issue on Magento 2.4-develop branch and provide missing steps if any if the issue is still reproducible.

Thank you.

**Comment by Assambra** [comment]
Thank you for working on this issue. What i see is on your sitemap.xml the default categorie are included this is not the issue i descriped but is this a new issue? You are sure that the sitemap.xml is correct for the rewrite picture?

**Comment by Assambra** [comment]
Step to reproduce the issue:
Create three categories -> Category1 Category2 Category3 each as subcategory from the last one.
Category1
    |->Category2
        |-->Category3

Create a product name it Product1 set the product category from this product to Subcategory3 (only)
Create the sitemap

**Comment by Assambra** [comment]
![1](https://user-images.githubusercontent.com/6276777/147650842-3cde603a-191a-4a95-ba9c-fc332532df5c.jpg)

**Comment by Assambra** [comment]
![2](https://user-images.githubusercontent.com/6276777/147651224-f9e07c7a-49be-4147-9eea-78a3a27a239b.jpg)

**Comment by Assambra** [comment]
![3](https://user-images.githubusercontent.com/6276777/147651460-c97b6d18-4587-4a0f-bd7c-85bb3632fd70.jpg)

**Comment by Assambra** [comment]
![4](https://user-images.githubusercontent.com/6276777/147652012-896381de-3635-41fc-b4c6-9e55057d6839.jpg)

**Comment by Assambra** [comment]
![5](https://user-images.githubusercontent.com/6276777/147653307-23e1d545-509d-47ae-85d0-4933a5bc24cc.jpg)

**Comment by engcom-November** [comment]
@magento give me 2.4-develop instance

**Comment by engcom-November** [comment]
Hi @Assambra, 
I re-verified the issue again on Magento 2.4-develop branch with below Steps followed:
1 - Admin - Stores - Configuration - Catalog - Catalog - Search Engine Optimization - Use Categories Path for Product URLs is set to YES and Save config
2. Created Categories as below:
![image](https://user-images.githubusercontent.com/60198231/147759091-853b1684-c9a0-4c4c-b353-143e23c543e6.png)

3. Created Product with only last sub category added as below:
![image](https://user-images.githubusercontent.com/60198231/147759021-22f5f0d3-bcb4-4372-b323-3100132cf644.png)

4. Generated Sitemap

Category path is not included for product url in sitemap no matter even if all categories are assigned to the product and this is an existing behavior of Magento but not a bug. 
![image](https://user-images.githubusercontent.com/60198231/147758981-eafed399-8cc1-4c9a-9520-6b054e3a8e33.png)

Created a product with all categories assigned and generated sitemap:
![image](https://user-images.githubusercontent.com/60198231/147759210-26f42012-32e3-468e-87a3-40e43e1c36e3.png)

@Assambra, I believe this is not a bug but the existing behavior of Magento. But as per your screenshot, Category path is included in sitemap.xml. Kindly confirm if you are using any third party extension in your setup. Kindly recheck the behavior on clean Magento 2.4-develop instance and provide missing steps if any if you are still facing any issues.
Thank you.

**Comment by Assambra** [comment]
I dont use any third party extension this is a clean magento 2.4.3-p1 installation on my server. 

What is normal behaviour? 
if i enable Use Categories Path for Product it should show in your case http://magento24.local/default/product9.html
Or it should show http://magento24.local/default/category1/category2/category3/product9.html
Please clearify me!

in any case something isnt work correct on your or my or booth magento installation.

your Sitemap.xml pictures show one time alle categories and other time only url + product. why?
I mean there is something incorrect and also i show you my result and have this problem on my production site and on the clean installed shop as i show you the pictures. 
My Sitemap creates in your example http://magento24.local/default/category1/category2/product9.html

What should it show?

greethings

**Comment by Assambra** [comment]
@magento give me 2.4-develop instance

**Comment by Assambra** [comment]
@magento give me 2.4-develop instance

**Comment by engcom-November** [comment]
As the behavior is different on Magento 2.4-develop branch, confirming the issue with updated description by adding additional information.

**Comment by DiVoronovich** [comment]
I decided to deal with this issue. The way @Assambra sees the solution to this issue was before. There was a call to such a method
![image](https://user-images.githubusercontent.com/92044471/151517931-9b21d9f0-efed-4c79-a86a-79fff028cfe1.png)

![image](https://user-images.githubusercontent.com/92044471/151517967-c3752356-b1aa-43c6-87d8-06e8afe55d13.png)

**And result was:**
product1.html
subcategory1/product1.html
subcategory1/subcategory2/product1.html
subcategory1/subcategory2/subcategory3/product1.html

I decided to figure out why they did it the way it is now. I found pull request https://github.com/magento/magento2/pull/29184 for the next issue https://github.com/magento/magento2/issues/23363 in which explains why it should be so.

**The main explanation for the irrelevance of this issue from https://github.com/magento/magento2/issues/23363 :**
>First problem
As far as I know, the sitemap.xml files' only purpose is for search engines to use as they can then more easily find all important links to index a webshop.
In this case, where Use Canonical Link Meta Tag For Products is set to Yes, we already indicate on product detail pages that search engines have to use the canonical url (which doesn't contain the category path) using a <meta> tag.
But after #23129 got merged, the sitemap files contain product url's with the category path and not the canonical url's. So search engines will first find the non-canonical url, then look at the source code of that page, then find the canonical url, and then visit that url and remove the first page they found in the sitemap file from their index. This means at least 2 requests are made for every product on your shop from search engines which might cause more traffic/load then needed.

>Second problem
Magento added a new option Generate "category/product" URL Rewrites (via MC-4244) which is enabled by default, but when you disable it, Magento will no longer store category paths for product url's in the url_rewrite table in the database, but still allows you to see url's with the category path on the frontend as those url's are then build dynamically.
Here it is expected that the sitemap files include the links being seen on the frontend even if they don't exist in the url_rewrite table, but the sitemap now uses the canonical links instead of the full url including the category path.

**Comment by engcom-Delta** [comment]
Hi @Assambra ,

Thanks for your reporting and collaboration.
We have re-verified the issue in Latest 2.4-develop instance and the issue is reproducible. Kindly refer the screenshots.

**Steps to reproduce**
1. Set Use Categories Path for Product URLs -> Yes
2. Catalog-->Categories -> Create Three categories each as Subcategory from the last one.
Default Categorie
|>Subcategory1
|-->Subcategory2
|_--->Subcategory3
3. Catalog-->Products -> Create a product and name it Product1, set the Product Category to Subcategory3
4. Marketing->Site Map -> Create a new sitemap and generate it
5. Observe Category path is not included for product url in sitemap no matter even if all categories are assigned to the product:
![image](https://github.com/user-attachments/assets/95dc0ca2-c9f9-4686-b533-648dc196d8c8)

Thanks.

**Comment by engcom-Bravo** [comment]
Adobe Commerce Engineering team started working on this issue. We will reach out to you if we need more information and you will get notified once the issue is fixed. Please leave comments for any further questions. Thank you!
